test/jdk/jdk/internal/reflect/CallerSensitive/CheckCSMs.java

```diff
@@ -111,170 +114,162 @@ public static void main(String[] args) throws Exception {
     private final Set<String> nonFinalCSMs = new ConcurrentSkipListSet<>();
     private final Map<String, Set<String>> csmWithCallerParameter = new ConcurrentHashMap<>();
 
-    private final ReferenceFinder finder;
     public CheckCSMs() {
-        this.finder = new ReferenceFinder(getFilter(), getVisitor());
         pool = Executors.newFixedThreadPool(numThreads);
-
     }
 
     public Set<String> run(Stream<Path> classes)
         throws IOException, InterruptedException, ExecutionException,
-               ConstantPoolException
+               IllegalArgumentException
     {
         classes.forEach(p -> pool.submit(getTask(p)));
         waitForCompletion();
         return nonFinalCSMs;
     }
 
+    private void check(ClassModel clazz) {
+        final String className = "jdk/internal/reflect/Reflection";
+        final String methodName = "getCallerClass";
+        boolean checkMethods = false;
+        for (var pe : clazz.constantPool()) {
+            if (pe instanceof MethodRefEntry ref
+                    && ref.owner().name().equalsString(className)
+                    && ref.name().equalsString(methodName)) {
+                checkMethods = true;
+            }
+        }
+
+        if (!checkMethods)
+            return;
 
-    private ReferenceFinder.Filter getFilter() {
-        final String classname = "jdk/internal/reflect/Reflection";
-        final String method = "getCallerClass";
-        return new ReferenceFinder.Filter() {
-            public boolean accept(ConstantPool cpool, CPRefInfo cpref) {
-                try {
-                    CONSTANT_NameAndType_info nat = cpref.getNameAndTypeInfo();
-                    return cpref.getClassName().equals(classname) && nat.getName().equals(method);
-                } catch (ConstantPoolException ex) {
-                    throw new RuntimeException(ex);
+        for (var method : clazz.methods()) {
+            var code = method.code().orElse(null);
+            if (code == null)
+                continue;
+
+            boolean needsCsm = false;
+            for (var element : code) {
```

Comment:
Scanning the instructions is a bit different approach than in the original test.